### ISSUE TITLE
Developers should have access to System.Text.Json's default internal converters

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/runtime#63791]
## Background and Motivation

Currently, the default json converters used for objects and collection types are internal to System.Text.Json. The library uses [hardcoded conventions](https://github.com/dotnet/runtime/blob/0b527a201f4c286f13ef8b829bbd1651bdad7a52/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L43-L48) to decide what default converters to apply to each type: for instance classes that implement `IEnumerable<T>` are automatically treated as collections, and classes that implement `IAsyncEnumerable<T>` can only be handled by the async-only IAsyncEnumerable converter. As of today, System.Text.Json offers no mechanism for users to override this convention, and we have received a number of questions from the community about this issue:

* https://github.com/dotnet/runtime/issues/1808
* https://github.com/dotnet/runtime/issues/61044

Currently the best available workaround is for users to either reimplement the default converters, or wrap/cast the values into a type that produces the desired serialization contract. At the same time, we offer no mechanism for users to tweak the serialization behaviour of the default converters other than using attribute annotations (to be ameliorated once #63686 has been implemented). Related issues:

* https://github.com/dotnet/runtime/issues/38514

## Proposal

This story proposes making the default internal converters public, so that users are free to opt in to specific converter strategies for their types. Consider the following example:

```C#
public class MyClass : IEnumerable<int>, IAsyncEnumerable<int>
{
}
```

By default, members of type `MyClass` are serialized as IAsyncEnumerables. Users can still force IEnumerable semantics if they cast the instance to `IEnumerable<int>`, however there is currently no way to serialize the type as an object. By exposing the default converter types, we can decorate our type definition with attributes:

```C#
[JsonConverter(typeof(JsonObjectConverter))] // serialize as object
// [JsonConverter(typeof(JsonCollectionConverter))] // serialize as collection
// [JsonConverter(typeof(JsonAsyncEnumerableConverter)] // serialize as IAsyncEnumerable
public class MyClass : IEnumerable<int>, IAsyncEnumerable<int>
{
}
```

Moreover, we should consider exposing parameterization on the converters so that users can configure the contract (for example what https://github.com/dotnet/runtime/issues/38514 is asking for).

## Progress

- [ ] Prototyping work
- [ ] API proposal
- [ ] Implementation & tests
- [ ] Documentation

**Comment by asik** [comment]
Related to #67361 , for me the ideal scenario would be that JsonConverter sees that it's an `IEnumerable<T>`, sees that it has a constructor that takes in an `IEnumerable<T>`, and uses that constructor.

**Comment by eiriktsarpalis** [comment]
> Related to #67361 , for me the ideal scenario would be that JsonConverter sees that it's an `IEnumerable<T>`, sees that it has a constructor that takes in an `IEnumerable<T>`, and uses that constructor.

Yes, that might be a good convention to honor. Similarly for dictionary types that accept `IEnumerable<KeyValuePair<TKey, TValue>>` parameters.

**Comment by eiriktsarpalis** [comment]
We likely won't have time to work on this before .NET 7 is done, moving to 8.0.0

**Comment by eiriktsarpalis** [comment]
Moving to Future.

**Comment by MooVC** [comment]
What is the suggested workaround in this case? I've a type that wraps an array (private field) that implements IEnumerable of that type. I would like it to be serialised as an object, using a public constructor that accepts an array. The type is intended to be immutable, so it would be counterintuitive to allow for collection overrides e.g. Add.

**Comment by eiriktsarpalis** [comment]
The only workaround for now is authoring a custom converter for the type, unfortunately.

**Comment by paul-michalik** [comment]
> What is the suggested workaround in this case? I've a type that wraps an array (private field) that implements IEnumerable of that type. I would like it to be serialised as an object, using a public constructor that accepts an array. The type is intended to be immutable, so it would be counterintuitive to allow for collection overrides e.g. Add.

I am currently solving this by deriving such types from `List<T>`. This provides the expected behavior (transparent serialization and deserialization as collection of `T`) but of course adds some risk if the collection should not be manipulated as, well `List<T>`. I mostly use this for internal types, so mostly harmless. 

I still do not understand what the problem is with implementing a default deserialization converter for types implementing e.g. `ICollection<T>`: the documentation says it all, this is exactly how it should behave:

<img width="541" alt="image" src="https://user-images.githubusercontent.com/10583150/221903867-5125e389-19aa-4158-9ab4-9fb7cea0f5fe.png">

**Comment by AbakumovAlexandr** [comment]
It will have been 3 years soon since this major limitation being in place. Is there any updates with this?

Without resolving this one, it's really a non-trivial task to implement even pretty basic requirements, for example, 'Deserialize all empty JSON arrays into `null`s instead of empty collections supporting all target .NET collection types'.

**Comment by eiriktsarpalis** [comment]
If there are any updates, they will be first posted on this issue. In the meantime, please continue upvoting the OP so that it can be prioritized in the future.

**Comment by mirek-rynly** [comment]
Really looking forward to this. We currently have to resort to fragile reflection to work around this limitation.

**Comment by ghost** [comment]
The `System.Text.Json.Serialization.Converters.ObjectDefaultConverter` should be for public usage, there is currently no way to for the `JsonConverterFactory` to return null, as it throws an exception when returning null


``` csharp
public override JsonConverter? CreateConverter(Type typeToConvert, JsonSerializerOptions options)
{
    //returning null throws exception in GetConverterInternal, even tho the method states that returning null is allowed
    return null;
}
```

Workaround is getting the Internal Converter through reflection

``` csharp
private static readonly Lazy<Type> _internalMicrosoftDefaultObjectConverterType = new(() => typeof(JsonConverter<>).Assembly.GetType("System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1")!.MakeGenericType(typeof(myTypeToConvert))!);

...

JsonConverter internalobjConverter = (JsonConverter)Activator.CreateInstance(_internalMicrosoftDefaultObjectConverterType.Value)!;

```

The reason why we need it, is because we have depending on different condition, different converter for this one object, and sometimes we must return the defauilt converter

**Comment by EdLichtman** [comment]
I'm currently building an auto parser, to convert our key store into a strongly typed object cache and have just run into the same issue, in that I basically have to specify every single available type for conversion or else I can't seem to use the default converters.

I'm very interested in this improvement.

**Comment by CoreDX9** [comment]
If I use a custom converter to convert custom types of implementing `IEnumerable<T>`, the exported JSON schema is incorrect. This became a significant flaw after `Microsoft.AspNetCore.OpenApi` became the preferred feature. Supporting the configuration of `Enumerable` as `Object` has become an urgent requirement.
